src/hooks/lifi/useShivaBuild.ts

```diff
@@ -0,0 +1,134 @@
+import { Address, encodeFunctionData, getContract, Hex } from "viem";
+import { DEFAULT_CHAINID } from "../../constants/chains";
+import { SHIVA_ADDRESS } from "../../constants/applications";
+import { useCallback } from "react";
+import { usePublicClient, useWalletClient } from "wagmi";
+import { ShivaABI } from "./abi/Shiva";
+
+const EIP712_DOMAIN = {
+  name: 'Shiva',
+  version: '1',
+  chainId: DEFAULT_CHAINID as number,
+  verifyingContract: SHIVA_ADDRESS[DEFAULT_CHAINID as number] as Address 
+};
+
+const EIP712_TYPES = {
+  BuildOnBehalfOf: [
+    { name: 'ovlMarket', type: 'address' },
+    { name: 'brokerId', type: 'uint32' },
+    { name: 'isLong', type: 'bool' },
+    { name: 'collateral', type: 'uint256' },
+    { name: 'leverage', type: 'uint256' },
+    { name: 'priceLimit', type: 'uint256' },
+    { name: 'owner', type: 'address' },
+    { name: 'deadline', type: 'uint48' },
+    { name: 'nonce', type: 'uint256' }
+  ]
+};
+
+export interface BuildOnBehalfOfParams {
+  marketAddress: Address;
+  brokerId?: number;
+  isLong: boolean;
+  collateral: bigint;
+  leverage: bigint;
+  priceLimit: bigint;
+  owner: Address; 
+  deadline: number; 
+}
+
+export const useShivaBuild = () => {
+  const publicClient = usePublicClient();
+  const { data: walletClient } = useWalletClient();
+
+  const getCurrentNonce = useCallback(
+  async (owner: Address): Promise<bigint> => {
+    if (!publicClient) throw new Error("publicClient is undefined");
+
+    const shivaContract = getContract({
+      address: SHIVA_ADDRESS[DEFAULT_CHAINID as number] as Address,
+      abi: ShivaABI,
+      client: publicClient,
+    });
+
+    let nonce = 0n;
+    while (await shivaContract.read.usedNonces([owner, nonce])) {
+      nonce++;
+    }
```

Comment:
Generate a random uint256 number and use it as a nonce. Retry if the nonce was taken before.
The current implementation is not scaleable.